### ISSUE TITLE
Revisit geospatial compound indexes

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#2855]
The proposal for geospatial indexes https://github.com/rethinkdb/rethinkdb/issues/2571 included an interface for compound geo indexes. However the proposal only permitted the geospatial component to be at the beginning of the index, which doesn't yield efficient performance in many use cases given the way that geospatial indexes are currently implemented.

The code currently in `next` does not permit compound geo indexes at all.

I'll add more details soon, as to where the problems lie and what our options are.

**Comment by thelinuxlich** [comment]
+1, without this any real use case for tables with multiple clients/departments/sectors/whatever is impracticable.

**Comment by adnaan** [comment]
I have a similar problem. I have a list of posts which are geotagged and I need to show the recent ones. 
My problem statement boils down to : 1) How do you fetch a set of a recent geotagged items in a polygon 2) How to do that efficiently?

As suggested by @danielmewes the following query fails: 

```
r.db('mydb').table('post').between(r.minval, r.maxval, {leftBound: "closed",rightBound:"open", index: "uid"}).orderBy(r.row('point')).distance(r.point(77.630939, 12.953997),{unit:'km'})

where uid is unix nano timestamp, an incrementing integer
```

with 

```
RqlRuntimeError: Expected type DATUM but found SELECTION:
SELECTION ON table(post) in:
r.db("mydb").table("post").between(r.minval, r.maxval, {leftBound: "closed", rightBound: "open", index: "uid"}).orderBy(r.row("point")).distance(r.point(77.630939, 12.953997), {unit: "km"})
```

I also tried `eqjoin` by maintaining another table of location,uid pairs:

```
r.db('mydb').table('postLocation').getNearest(r.point(77.630939, 12.953997),{index:'point',maxDist:20,unit:'km'}).eqJoin(r.row("doc")("uid"), r.db('mydb').table("post"))
```

failed with 

No results returned.

Met with some success using `filter`

```
r.db('mydb').table('post').getNearest(r.point(77.630939, 12.953997),{index:'point',maxDist:20,unit:'km'}).filter(r.row('doc')('uid').lt(1446971060546369000)).limit(10)

where 1446971060546369000 is uid
```

But as you would expect the result is a set of posts ordered by `dist` not `uid` and hence it is not recent.

I would have expected something like [Mongodb compound indexes](https://docs.mongodb.org/manual/tutorial/build-a-2dsphere-index/#create-a-compound-index-with-2dsphere-index-key). So the above query would be like:

```
r.db('mydb').table('post').getNearest(r.point(77.630939, 12.953997),{indexes:{'point','uid'},maxDist:20,unit:'km'}).filter(r.row('doc')('uid').lt(1446971060546369000)).limit(10) 
````

or even better 

```
r.db('mydb').table('post').getNearest(r.point(77.630939, 12.953997),{indexes:{'point','uid'},maxDist:20,unit:'km'}).between(r.minval, r.maxval, {leftBound: "closed",rightBound:"open", index: "uid"})orderBy({index:r.desc('uid')}).limit(10)
```

In short, I am not sure how to move forward from here. Ideas?

**Comment by adnaan** [comment]
Forgot to mention this.

The best solution I have right now is the one suggested above using `getIntersecting`

```
r.db('mydb').table('post').getIntersecting(r.circle(r.point(77.630939, 12.953997),20,{unit:'km'}),{index:'area'}).filter(r.row('uid').lt(1446979942222195200))
```

I have opted to go with this for now. Is there a proposal plan which I could follow for implementation/discussion of geospatial compound queries.

**Comment by danielmewes** [comment]
@adnaan I think the first query has the parentheses set wrong:
```
r.db('mydb').table('post').between(r.minval, r.maxval, {leftBound: "closed",rightBound:"open", index: "uid"}).orderBy(r.row('point').distance(r.point(77.630939, 12.953997),{unit:'km'}))
```

Note that the `.distance` must be inside the `orderBy`.

I suspect that the `getIntersecting` followed by a `filter` is probably the best option available right now. Unless the `filter` is the more selective part, in which case the `between.orderBy(...distance)` might give you better performance.

**Comment by adnaan** [comment]
@danielmewes that sounds reasonable.

**Comment by kriestof** [comment]
In my opinion it would be really cool if rethinkdb would be able to merge lines with timestamps to create index.

Consider table with objects like:
{"ts": [array], "path": r.line(...), ...}

I would like to create index like:
r.table(table_name).indexCreate("positionTime", function(row) {
    return r.map(row("ts"), row("path"), function(a, b) {
        return [a, b]
    }, {multi: true, geo: true})
}).run()

As a way around right now one can use index on ts to prefilter objects with timestamp within the range and then make more advanced filtering by both timestamp and position. As long as number of objects within timestamp and number of queries is reasonable it should work out. 

I know a lot of things changed since this issue was created but maybe at some point in the feature this info would be helpful.
